File: Wire-iOS/Sources/UserInterface/Components/Transitions/VerticalTransition.m
// 
// Wire
// Copyright (C) 2016 Wire Swiss GmbH
// 
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program. If not, see http://www.gnu.org/licenses/.
// 


#import "VerticalTransition.h"
#import "UIView+WR_ExtendedBlockAnimations.h"

@interface VerticalTransition ()

@property (nonatomic) CGFloat offset;

@end

@implementation VerticalTransition

- (id)initWithOffset:(CGFloat)offset
{
    self = [self init];
    
    if (self) {
        _offset = offset;
    }
    
    return self;
}

- (NSTimeInterval)transitionDuration:(id<UIViewControllerContextTransitioning>)transitionContext
{
    return 0.55f;
}

- (void)animateTransition:(id<UIViewControllerContextTransitioning>)transitionContext
{
    UIView *fromView = [transitionContext viewForKey:UITransitionContextFromViewKey];
    UIView *toView = [transitionContext viewForKey:UITransitionContextToViewKey];
    UIView *containerView = [transitionContext containerView];
    
    toView.frame = [transitionContext finalFrameForViewController:[transitionContext viewControllerForKey:UITransitionContextToViewControllerKey]];
    fromView.frame = [transitionContext initialFrameForViewController:[transitionContext viewControllerForKey:UITransitionContextFromViewControllerKey]];
    
    [containerView addSubview:toView];
    
    if (! [transitionContext isAnimated]) {
        [transitionContext completeTransition:YES];
        return;
    }
    
    CGRect finalRect = [transitionContext finalFrameForViewController:[transitionContext viewControllerForKey:UITransitionContextToViewControllerKey]];
    
    double sign = copysign(1.0f, self.offset);
    CGAffineTransform toTransfrom = CGAffineTransformMakeTranslation(0, -self.offset);
    CGAffineTransform fromTransform = CGAffineTransformMakeTranslation(0, sign * (finalRect.size.height - fabs(self.offset)));
    
    fromView.transform = fromTransform;
    toView.transform = toTransfrom;
    
    NSArray <UIView *> *viewsToHide = [self.dataSource viewsToHideDuringVerticalTransition:self];
    
    [viewsToHide enumerateObjectsUsingBlock:^(UIView * _Nonnull obj, NSUInteger idx, BOOL * _Nonnull stop) {
        obj.hidden = YES;
    }];
    
    [UIView wr_animateWithEasing:WREasingFunctionEaseOutExpo duration:[self transitionDuration:transitionContext] delay:0 animations:^{
        fromView.transform = CGAffineTransformMakeTranslation(0.0f, sign * finalRect.size.height);
        toView.transform = CGAffineTransformIdentity;
    } completion:^(BOOL finished) {
        fromView.transform = CGAffineTransformIdentity;
        [viewsToHide enumerateObjectsUsingBlock:^(UIView * _Nonnull obj, NSUInteger idx, BOOL * _Nonnull stop) {
            obj.hidden = NO;
        }];
        [transitionContext completeTransition:YES];
    }];
}

@end

File: wire-ios/Wire-iOS/Sources/UserInterface/Components/Transitions/VerticalTransition.swift
//
// Wire
// Copyright (C) 2024 Wire Swiss GmbH
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program. If not, see http://www.gnu.org/licenses/.
//

import UIKit

// MARK: - VerticalTransitionDataSource

protocol VerticalTransitionDataSource: NSObject {
    func viewsToHideDuringVerticalTransition(transition: VerticalTransition) -> [UIView]
}

// MARK: - VerticalTransition

final class VerticalTransition: NSObject, UIViewControllerAnimatedTransitioning {
    // MARK: Lifecycle

    init(offset: CGFloat) {
        self.offset = offset

        super.init()
    }

    // MARK: Internal

    weak var dataSource: VerticalTransitionDataSource?

    func transitionDuration(using transitionContext: UIViewControllerContextTransitioning?) -> TimeInterval {
        0.55
    }

    func animateTransition(using transitionContext: UIViewControllerContextTransitioning) {
        let containerView = transitionContext.containerView

        guard let toView = transitionContext.toView,
              let toViewController = transitionContext.toViewController else { return }

        guard let fromView = transitionContext.fromView,
              let fromViewController = transitionContext.fromViewController else { return }

        fromView.frame = transitionContext.initialFrame(for: fromViewController)

        containerView.addSubview(toView)

        if !transitionContext.isAnimated {
            transitionContext.completeTransition(true)
            return
        }

        containerView.layoutIfNeeded()

        let sign = copysign(1.0, offset)
        let finalRect = transitionContext.finalFrame(for: toViewController)
        let toTransfrom = CGAffineTransform(translationX: 0, y: -offset)
        let fromTransform = CGAffineTransform(translationX: 0, y: sign * (finalRect.size.height - abs(offset)))

        toView.transform = toTransfrom
        fromView.transform = fromTransform

        if let viewsToHide = dataSource?.viewsToHideDuringVerticalTransition(transition: self) {
            viewsToHide.forEach { $0.isHidden = true }
        }

        UIView.animate(
            easing: EasingFunction.easeOutExpo,
            duration: transitionDuration(using: transitionContext),
            animations: {
                fromView.transform = CGAffineTransform(translationX: 0.0, y: sign * finalRect.size.height)
                toView.transform = CGAffineTransform.identity
            },
            completion: { _ in
                fromView.transform = CGAffineTransform.identity
                if let viewsToHide = self.dataSource?.viewsToHideDuringVerticalTransition(transition: self) {
                    viewsToHide.forEach { $0.isHidden = false }
                }

                transitionContext.completeTransition(true)
            }
        )
    }

    // MARK: Private

    private let offset: CGFloat
}
